Assign each RpcAgent a unique ID, and use ID for sending RPC messages.

mrshenli opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #25169 Adding RRef as return value for builtin operators
* #24260 Move worker name collection code from Python to C++
* **#24195 Assign each RpcAgent a unique ID, and use ID for sending RPC messages.**

Summary: It is not efficient to use a string destination name in every
send. Moreover, when we add RRef later, RpcAgent will frequently check
RRef ownership. It will be slow as well if we have to go though string
comparison every time. This commit assigns each RpcAgent a unique
integer ID. In the Python send API, applications can provide either
destination name or id. If it is a string name, it will be converted to
id by calling the get_id(workerName) API.

Differential Revision: [D16770241](https://our.internmc.facebook.com/intern/diff/D16770241)

mrshenli committed:
Assign each RpcAgent a unique ID, and use ID for sending RPC messages.

Summary: It is not efficient to use a string destination name in every
send. Moreover, when we add RRef later, RpcAgent will frequently check
RRef ownership. It will be slow as well if we have to go though string
comparison every time. This commit assigns each RpcAgent a unique
integer ID. In the Python send API, applications can provide either
destination name or id. If it is a string name, it will be converted to
id by calling the get_id(workerName) API.

pietern commented on test/test_rpc.py:
```diff
@@ -43,6 +43,15 @@ def test_add(self):
+    def test_add_with_id(self):
+        n = self.rank + 1
+        dstRank = n % self.world_size
+        workder_id = dist.get_id('worker{}'.format(dstRank))
```
Inconsistent snake and camel casing.

pietern commented on torch/csrc/distributed/rpc/ProcessGroupAgent.cpp:
```diff
@@ -73,14 +73,18 @@ ProcessGroupAgent::ProcessGroupAgent(
   listenerThread_ = std::thread(&ProcessGroupAgent::listenLoop, this);
 }
 
+uint64_t ProcessGroupAgent::getId(const std::string& workerName) {
```
Here it's an uint64_t but I can see this being a slightly richer structure some time in the future. Could you add a type alias on the RPCAgent class for this? E.g. `using worker_id_t = uint64_t`.

If it won't be a richer structure then at least it serves as a bit of documentation.

pietern commented on torch/csrc/distributed/rpc/ProcessGroupAgent.cpp:
```diff
@@ -73,14 +73,18 @@ ProcessGroupAgent::ProcessGroupAgent(
 }
 
+uint64_t ProcessGroupAgent::getId(const std::string& workerName) {
+  auto idIter = nameMap_.find(workerName);
```
const

pietern commented on torch/csrc/distributed/rpc/init.cpp:
```diff
@@ -49,6 +49,9 @@ PyObject* rpc_init(PyObject* /* unused */) {
                         std::shared_ptr<::c10d::ProcessGroup>>())
+          .def("get_id",
+               &ProcessGroupAgent::getId,
+               py::call_guard<py::gil_scoped_release>())
```
This reads as if `get_id` returns the ID of the agent itself.

The ID should be globally consistent and be serializable, so that it can be passed to a remote function as indirection (e.g. A tells B to call C). Then, we will need another function that returns the ID of the agent itself, which would likely be called `get_id`. What do you think of renaming this to `get_worker_id`, to indicate it takes an argument and does not necessarily apply to the local instance?

mrshenli replied:
good point! Will make the change.

mrshenli committed:
Update on "Assign each RpcAgent a unique ID, and use ID for sending RPC messages."


Summary: It is not efficient to use a string destination name in every
send. Moreover, when we add RRef later, RpcAgent will frequently check
RRef ownership. It will be slow as well if we have to go though string
comparison every time. This commit assigns each RpcAgent a unique
integer ID. In the Python send API, applications can provide either
destination name or id. If it is a string name, it will be converted to
id by calling the get_id(workerName) API.

Differential Revision: [D16770241](https://our.internmc.facebook.com/intern/diff/D16770241)

pritamdamania87 commented on torch/distributed/rpc.py:
```diff
@@ -96,6 +99,14 @@ def init_rpc(name, backend='pg'):
 
 
+def get_id(workerName=None):
+    _check_initialized()
```
nit: A decorator would be a nice way to do this for all functions.

mrshenli replied:
Good point, let me add that.

xush6528 replied:
Python always uses snake case
```
def get_id(worker_name=None):
```

pritamdamania87 commented on torch/distributed/rpc.py:
```diff
@@ -96,6 +99,14 @@ def init_rpc(name, backend='pg'):
         raise RuntimeError("Unrecognized RPC backend ", backend)
 
 
+def get_id(workerName=None):
```
nit: call this `get_worker_id` and add docs?

xush6528 replied:
agree. I'd prefer always make names descriptive.
Later, we will also have ref_id, fork_id.

pritamdamania87 commented on torch/csrc/distributed/rpc/RpcAgent.h:
```diff
@@ -7,58 +7,66 @@ namespace torch {
 namespace distributed {
 namespace rpc {
 
+using worker_id_t = uint64_t;
```
Was wondering if we should make this int16_t similar to the distributed autograd context PR? Do we expect to have more than 65k workers? I'm assuming ranks will always be contiguous integers starting from 0. Is this assumption fair?

mrshenli replied:
Assumption makes sense. If we truncate ranks, I think we should also apply explicit checks, as we break promises implied by the `ProcessGroup` API.

pritamdamania87 replied:
Is this `worker_id` part of our model parallel API and not really the `ProcessGroup` API? I feel we can truncate and add checks here to enforce this.

xush6528 commented on torch/csrc/distributed/rpc/ProcessGroupAgent.cpp:
```diff
@@ -73,14 +73,22 @@ ProcessGroupAgent::ProcessGroupAgent(
   listenerThread_ = std::thread(&ProcessGroupAgent::listenLoop, this);
 }
 
+worker_id_t ProcessGroupAgent::getId() {
```
Can we call this method getWorkerId() too? just does not take argument.

xush6528 replied:
Oh. Checked Pieter's comment. Seems he has a different naming convention.

mrshenli replied:
> Will this bring challenge to pybind ?

I should be able to do `py::overload_cast` to merge these two functions into one.

xush6528 commented on torch/csrc/distributed/rpc/init.cpp:
```diff
@@ -58,11 +64,11 @@ PyObject* rpc_init(PyObject* /* unused */) {
   module.def("invoke_rpc", [](
       RpcAgent& agent,
-      const std::string& dstName,
+      uint64_t dst,
```
worker_id_t ?

xush6528 commented on torch/csrc/distributed/rpc/python_functions.cpp:
```diff
@@ -20,7 +20,7 @@ py::object to_py_obj(const Message& message) {
 std::shared_ptr<FutureMessage> py_rpc(
     RpcAgent& agent,
-    const std::string& dstName,
+    uint64_t dst,
```
same

xush6528 commented on torch/csrc/distributed/rpc/python_functions.h:
```diff
@@ -18,7 +18,7 @@ py::object to_py_obj(const Message& message);
 std::shared_ptr<FutureMessage> py_rpc(
     RpcAgent& agent,
-    const std::string& dstName,
+    uint64_t dst,
```
same

xush6528 commented on torch/csrc/distributed/rpc/ProcessGroupAgent.cpp:
```diff
@@ -117,14 +125,9 @@ void ProcessGroupAgent::sync() {
-  TORCH_CHECK(dstRank != pg_->getRank(), "ProcessGroupAgent does not support "
-    "making RPC calls to self.")
+    worker_id_t to, Message&& message) {
+  TORCH_CHECK(to != (worker_id_t)pg_->getRank(),
```
Should we keep checking `to < pg_->world_size_`?

mrshenli replied:
Yes, let me add that

mrshenli committed:
Update on "Assign each RpcAgent a unique ID, and use ID for sending RPC messages."


Summary: It is not efficient to use a string destination name in every
send. Moreover, when we add RRef later, RpcAgent will frequently check
RRef ownership. It will be slow as well if we have to go though string
comparison every time. This commit assigns each RpcAgent a unique
integer ID. In the Python send API, applications can provide either
destination name or id. If it is a string name, it will be converted to
id by calling the get_id(workerName) API.

Differential Revision: [D16770241](https://our.internmc.facebook.com/intern/diff/D16770241)

apaszke reviewed (changes requested):
Can we at least make the worker id a custom type on the Python side to make sure people will not e.g. hardcode the worker IDs? In general what you're doing here seems extremely similar to `torch.device` which can be constructed from a string, but has a much more efficient internal representation which can be reused.

mrshenli committed:
Update on "Assign each RpcAgent a unique ID, and use ID for sending RPC messages."


Summary: It is not efficient to use a string destination name in every
send. Moreover, when we add RRef later, RpcAgent will frequently check
RRef ownership. It will be slow as well if we have to go though string
comparison every time. This commit assigns each RpcAgent a unique
integer ID. In the Python send API, applications can provide either
destination name or id. If it is a string name, it will be converted to
id by calling the get_id(workerName) API.

Differential Revision: [D16770241](https://our.internmc.facebook.com/intern/diff/D16770241)

mrshenli committed:
Update on "Assign each RpcAgent a unique ID, and use ID for sending RPC messages."


Summary: It is not efficient to use a string destination name in every
send. Moreover, when we add RRef later, RpcAgent will frequently check
RRef ownership. It will be slow as well if we have to go though string
comparison every time. This commit assigns each RpcAgent a unique
integer ID. In the Python send API, applications can provide either
destination name or id. If it is a string name, it will be converted to
id by calling the get_id(workerName) API.

Differential Revision: [D16770241](https://our.internmc.facebook.com/intern/diff/D16770241)

mrshenli committed:
Update on "Assign each RpcAgent a unique ID, and use ID for sending RPC messages."


Summary: It is not efficient to use a string destination name in every
send. Moreover, when we add RRef later, RpcAgent will frequently check
RRef ownership. It will be slow as well if we have to go though string
comparison every time. This commit assigns each RpcAgent a unique
integer ID. In the Python send API, applications can provide either
destination name or id. If it is a string name, it will be converted to
id by calling the get_id(workerName) API.

Differential Revision: [D16770241](https://our.internmc.facebook.com/intern/diff/D16770241)

mrshenli committed:
Update on "Assign each RpcAgent a unique ID, and use ID for sending RPC messages."


Summary: It is not efficient to use a string destination name in every
send. Moreover, when we add RRef later, RpcAgent will frequently check
RRef ownership. It will be slow as well if we have to go though string
comparison every time. This commit assigns each RpcAgent a unique
integer ID. In the Python send API, applications can provide either
destination name or id. If it is a string name, it will be converted to
id by calling the get_id(workerName) API.

Differential Revision: [D16770241](https://our.internmc.facebook.com/intern/diff/D16770241)

xush6528 reviewed (approved):
All look good to me now

mrshenli committed:
Update on "Assign each RpcAgent a unique ID, and use ID for sending RPC messages."


Summary: It is not efficient to use a string destination name in every
send. Moreover, when we add RRef later, RpcAgent will frequently check
RRef ownership. It will be slow as well if we have to go though string
comparison every time. This commit assigns each RpcAgent a unique
integer ID. In the Python send API, applications can provide either
destination name or id. If it is a string name, it will be converted to
id by calling the get_id(workerName) API.

Differential Revision: [D16770241](https://our.internmc.facebook.com/intern/diff/D16770241)

mrshenli commented:
Test failures are irrelevant. If there is no more change requests, I will rebase and land this later today. @apaszke  @pietern @pritamdamania87  

```
Aug 28 19:21:11 I0828 19:21:10.697399   164 operator_test.cc:101] Exception [enforce fail at operator.cc:273] op. Cannot create operator of type 'JustTypeCPUOnly' on the device 'CUDA'. Verify that implementation for the corresponding device exist. It might also happen if the binary is not linked with the operator implementation code. If Python frontend is used it might happen if dyndep.InitOpsLibrary call is missing. Operator def: type: "JustTypeCPUOnly" device_option { device_type: 1 }
```

```
20:10:57 [  FAILED  ] DistributionsTest.TestPhiloxIncrementSmallMultinomialTensor (2430 ms)
```

pritamdamania87 commented on torch/csrc/distributed/rpc/process_group_agent.cpp:
```diff
@@ -69,21 +72,37 @@ ProcessGroupAgent::ProcessGroupAgent(
+const WorkerId& ProcessGroupAgent::getWorkerId(
+    const std::string& workerName) const {
+  const auto idIter = nameMap_.find(workerName);
+  TORCH_CHECK(idIter != nameMap_.end(),
```
Would be good to add a few unit tests that trigger errors like these, ex:

1. Pass an invalid worker name.
2. Pass invalid worker id.
3. Pass worker name as self.
4. Pass worker id as self.

mrshenli replied:
Yes, will add those, thanks!

mrshenli committed:
Update on "Assign each RpcAgent a unique ID, and use ID for sending RPC messages."


Summary: It is not efficient to use a string destination name in every
send. Moreover, when we add RRef later, RpcAgent will frequently check
RRef ownership. It will be slow as well if we have to go though string
comparison every time. This commit assigns each RpcAgent a unique
integer ID. In the Python send API, applications can provide either
destination name or id. If it is a string name, it will be converted to
id by calling the get_id(workerName) API.

Differential Revision: [D16770241](https://our.internmc.facebook.com/intern/diff/D16770241)

zhaojuanmao commented:
@mrshenli  #24875 also has getWorkerId() to return int16_t

mrshenli commented:
@zhaojuanmao Thanks for the reminder, I addressed that during rebase. Landing now.
